extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProvider.java

```diff
@@ -119,56 +128,87 @@ public SeekableByteChannel getWriteChannel(@NotNull final Path path, final boole
+                if (currentIndex < currentBatch.size()) {
+                    return true;
+                }
+                if (fetchedAllURIs) {
+                    return false;
+                }
+                try {
+                    fetchNextBatch();
+                } catch (final IOException e) {
+                    throw new UncheckedDeephavenException("Failed to fetch next batch of URIs from S3", e);
+                }
+                return currentIndex < currentBatch.size();
+            }
+
+            @Override
+            public URI next() {
+                if (!hasNext()) {
+                    throw new NoSuchElementException("No more URIs available in the directory");
+                }
+                return currentBatch.get(currentIndex++);
+            }
+
+            private void fetchNextBatch() throws IOException {
+                currentIndex = 0;
+                currentBatch.clear();
+                final S3Uri s3DirectoryURI = s3AsyncClient.utilities().parseUri(directory);
+                final String bucketName = s3DirectoryURI.bucket().orElseThrow();
+                final String directoryKey = s3DirectoryURI.key().orElseThrow();
+                final ListObjectsV2Request.Builder requestBuilder = ListObjectsV2Request.builder()
+                        .bucket(bucketName)
+                        .prefix(s3DirectoryURI.key().orElseThrow());
+                if (!isRecursive) {
+                    // Add a delimiter to the request if we don't want to fetch all files recursively
+                    requestBuilder.delimiter("/");
+                }
+                final long readTimeoutNanos = s3Instructions.readTimeout().toNanos();
+                final ListObjectsV2Response response;
+                final ListObjectsV2Request request = requestBuilder.continuationToken(continuationToken).build();
+                try {
+                    response = s3AsyncClient.listObjectsV2(request).get(readTimeoutNanos, TimeUnit.NANOSECONDS);
+                } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+                    throw handleS3Exception(e, String.format("fetching list of files in directory %s", directory),
+                            s3Instructions);
+                }
+                response.contents().stream()
+                        .filter(s3Object -> !s3Object.key().equals(directoryKey))
+                        .map(s3Object -> URI.create("s3://" + bucketName + "/" + s3Object.key()))
+                        .forEach(currentBatch::add);
+                // If the response is truncated, fetch the next page using the continuation token from the response
+                // Usually the response is truncated when there are more than 1000 objects in the bucket
+                // TODO Test this
+                fetchedAllURIs = !response.isTruncated();
+                continuationToken = response.nextContinuationToken();
```

Comment:
I think we could do away w/ `fetchedAllURIs` and just rely on if `continuationToken` is `null`?